saw-core/src/SAWCore/Term/CtxTerm.hs

```diff
@@ -88,263 +47,42 @@ import SAWCore.Term.Functor
--- is.
-newtype CtxTerm (ctx :: Ctx Type) (a :: Type) = CtxTerm Term
-
--- | Convert a 'CtxTerm' to an untyped term. This is "unsafe" because it throws
--- away our typing information.
-unCtxTermUnsafe :: CtxTerm ctx a -> Term
-unCtxTermUnsafe (CtxTerm t) = t
-
--- | Because we cannot capture the SAW type system in Haskell, sometimes we have
--- to cast our terms. We try not to use this very often, and we only allow
--- casting the output type, not the context, since the latter could screw up our
--- deBruijn indices.
-castCtxTerm :: Proxy a -> Proxy b -> CtxTerm ctx a -> CtxTerm ctx b
-castCtxTerm _ _ (CtxTerm t) = CtxTerm t
-
--- | Build a term in the empty context
-mkClosedTerm :: Term -> CtxTerm EmptyCtx a
-mkClosedTerm = CtxTerm
-
--- | Build a term to represent a type in the empty context
-mkClosedTyp :: Term -> CtxTerm EmptyCtx (Typ a)
-mkClosedTyp = mkClosedTerm
-
--- | Take a term out of the empty context
-elimClosedTerm :: CtxTerm EmptyCtx a -> Term
-elimClosedTerm (CtxTerm t) = t
-
--- | Existentially quantify over the "type" of an object
-data ExistsTp tp ctx = forall a. ExistsTp (tp ctx a)
-
--- | Build a 'Bindings' list from a list of variable names and types, assuming
--- that each variable is free in the remaining types and that @ctx@ describes
--- the ambient context of the top-level type in the context. Note that nothing
--- here ensures that the Haskell-level types used to "represent" the bindings
--- created by this function have anything to do with the actual SAW types, but
--- the Haskell type system is not powerful enough to represent all the SAW types
--- anyway, and any code that consumes this 'Bindings' list cannot know that
--- anyway. See also the comments for 'CtxTerm'.
-ctxBindingsOfTerms :: [(LocalName, Term)] -> ExistsTp (Bindings CtxTerm) ctx
-ctxBindingsOfTerms [] = ExistsTp NoBind
-ctxBindingsOfTerms ((x,tp):ctx) =
-  case ctxBindingsOfTerms ctx of
-    ExistsTp rest -> ExistsTp (Bind x (CtxTerm tp) rest)
-
--- | A dummy unit type that takes in a context
-data CtxUnit ctx a = CtxUnit
-
--- | An 'Either' type relative to a context and type
-newtype CtxEither f g ctx a = CtxEither (Either (f ctx a) (g ctx a))
-
--- | A list of terms in a given context
-data CtxTerms ctx as where
-  CtxTermsNil :: CtxTerms ctx '[]
-  CtxTermsCons :: CtxTerm ctx a -> CtxTerms ctx as -> CtxTerms ctx (a ': as)
-
--- | A list of terms in a given context, stored "inside-out"
-data CtxTermsCtx ctx term_ctx where
-  CtxTermsCtxNil :: CtxTermsCtx ctx EmptyCtx
-  CtxTermsCtxCons :: CtxTermsCtx ctx as -> CtxTerm ctx a ->
-                     CtxTermsCtx ctx (as ::> a)
-
-{-
--- | Get the head and tail of a non-empty 'CtxTerms' list
-ctxTermsHeadTail :: CtxTerms ctx (a ': as) -> (CtxTerm ctx a, CtxTerms ctx as)
-ctxTermsHeadTail (CtxTermsCons a as) = (a, as)
--}
+-- | Get the head and tail of a non-empty '[Term]' list
+ctxTermsCtxHeadTail :: [Term] -> ([Term], Term)
+ctxTermsCtxHeadTail as = (init as, last as)
```

Comment:
After the changes in this PR, this turns from a total function to a partial function. This is OK, I think, but it might be worth advertising in the Haddocks that this function will throw an exception if you give it an empty list.

In addition, it might be worth redefining this function in terms of [`unsnoc`](https://hackage-content.haskell.org/package/extra-1.8/docs/Data-List-Extra.html#v:unsnoc) so that we can throw a specific error message instead of a generic one like `Prelude.init: empty list`.

Reply:
Fixed in 94459a478425b5ea589e5817f60ef1473f4db7ef.